# SQL PROJECT. Book Market Analysis.

# Task Description

- To find the number of books released after January 1, 2000.
- To find the number of user reviews and the average rating for each book.
- To identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- To identify the author with the highest average book rating (look only at books with at least 50 ratings).
- To find the average number of text reviews among users who rated more than 50 books.

In [1]:
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student',        
             'pwd': 'Sdf4$2;d-d30pp',
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,             
             'db': 'data-analyst-final-project-db'}      

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Data Study

In [2]:
books = pd.io.sql.read_sql("SELECT * FROM books", con = engine)

In [3]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [5]:
books = books.dropna()
books.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 54.7+ KB


In [6]:
authors = pd.io.sql.read_sql("SELECT * FROM  authors", con = engine)

In [7]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [8]:
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [9]:
authors = authors.dropna()
authors.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 14.9+ KB


In [10]:
ratings = pd.io.sql.read_sql("SELECT * FROM  ratings", con = engine)

In [11]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [12]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [13]:
ratings = ratings.dropna()
ratings.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 252.2+ KB


In [14]:
reviews = pd.io.sql.read_sql("SELECT * FROM reviews", con = engine)

In [15]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [16]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [17]:
reviews = reviews.dropna()
reviews.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 109.1+ KB


In [18]:
publishers = pd.io.sql.read_sql("SELECT * FROM publishers", con = engine)

In [19]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [20]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [21]:
publishers = publishers.dropna()
publishers.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 8.0+ KB


## Finding the number of books released after January 1, 2000.

In [38]:
books_after_jan1_2000 = (pd.io.sql.read_sql("SELECT COUNT(book_id) AS books_count FROM books WHERE publication_date >= '2000-01-01'", con = engine))
books_after_jan1_2000

books_count
0          821

There are 821 books published after January 1, 2020.

## Finding the number of user reviews and the average rating for each book.

In [23]:
books_reviews = pd.io.sql.read_sql("SELECT title, book_id, COUNT(review_id) AS reviews_count FROM books JOIN reviews USING (book_id) GROUP BY (book_id) ORDER BY (reviews_count) DESC", con = engine)
books_reviews.head(10)

title  book_id  reviews_count
0                            Twilight (Twilight  #1)      948              7
1                                     The Book Thief      656              6
2                          Outlander (Outlander  #1)      497              6
3                                Water for Elephants      963              6
4             The Da Vinci Code (Robert Langdon  #2)      696              6
5  The Curious Incident of the Dog in the Night-Time      695              6
6                                           The Road      854              6
7                                      The Alchemist      627              6
8                                    Eat  Pray  Love      207              6
9                                   The Glass Castle      734              6

#### SubConclusions:
- The biggest number of user reviews for a book is 7, and it belongs to Twilligt (#1);
- The lowest number of user reviews for a book is 1.

In [24]:
books_ratings = pd.io.sql.read_sql("SELECT title, book_id, AVG(rating) AS avg_rating FROM books JOIN ratings USING (book_id) GROUP BY (book_id) ORDER BY (avg_rating) DESC", con = engine)
books_ratings.head(10) 

title  book_id  avg_rating
0               Pop Goes the Weasel (Alex Cross  #5)      518         5.0
1  The Ghost Map: The Story of London's Most Terr...      732         5.0
2  In the Hand of the Goddess (Song of the Liones...      347         5.0
3                           Tai-Pan (Asian Saga  #2)      610         5.0
4  How to Be a Domestic Goddess: Baking and the A...      330         5.0
5           Evil Under the Sun (Hercule Poirot  #24)      229         5.0
6                                          Geek Love      275         5.0
7            School's Out—Forever (Maximum Ride  #2)      553         5.0
8                      Emily of New Moon (Emily  #1)      213         5.0
9                                              Women      993         5.0

In [25]:
books_ratings.tail(10) 

title  book_id  avg_rating
990               Tsubasa: RESERVoir CHRoNiCLE  Vol. 1      943    2.500000
991                                     Prep : A Novel      525    2.500000
992          Neither Here nor There: Travels in Europe      470    2.500000
993                                  The Mermaid Chair      794    2.333333
994                             The Kitchen God's Wife      772    2.333333
995  The World Is Flat: A Brief History of the Twen...      915    2.250000
996                                              Junky      371    2.000000
997                  His Excellency: George Washington      316    2.000000
998                                      Drowning Ruth      202    2.000000
999                               Harvesting the Heart      303    1.500000

#### SubConclusions
- The highest average rating is 5.0;
- The lowest average rating (1.5) belongs to "Harvesting the Heart".

Let's merge findings into one dataframe

In [26]:
books_ratings_reviews = pd.io.sql.read_sql("WITH books_ratings AS(SELECT title, book_id, AVG(rating) AS avg_rating FROM books JOIN ratings USING (book_id) GROUP BY (book_id)), books_reviews AS (SELECT title, book_id, COUNT(review_id) AS reviews_count FROM books JOIN reviews USING (book_id) GROUP BY (book_id)) SELECT title, avg_rating, reviews_count FROM books_ratings JOIN books_reviews USING (title) ORDER BY (reviews_count) DESC", con = engine)
books_ratings_reviews.head()

title  avg_rating  reviews_count
0  Twilight (Twilight  #1)    3.662500              7
1                 The Road    3.772727              6
2           The Book Thief    4.264151              6
3      Water for Elephants    3.977273              6
4         The Glass Castle    4.206897              6

## Identifying the publisher that has released the greatest number of books with more than 50 pages.

In [27]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [28]:
book_publisher = (pd.io.sql.read_sql("SELECT publisher_id, publisher, COUNT(book_id) AS books_count FROM publishers JOIN books USING (publisher_id) WHERE num_pages > 50 GROUP BY (publisher_id) ORDER BY (books_count) DESC LIMIT 1", con = engine))
book_publisher.head()

publisher_id      publisher  books_count
0           212  Penguin Books           42

### Conclusion:
 - Penguin Books has released the greatest number of books (42) of more than 50 pages.

## Identifying the author with the highest average book rating (look for books with at least 50 ratings).

In [29]:
books_ratings.head()

title  book_id  avg_rating
0               Pop Goes the Weasel (Alex Cross  #5)      518         5.0
1  The Ghost Map: The Story of London's Most Terr...      732         5.0
2  In the Hand of the Goddess (Song of the Liones...      347         5.0
3                           Tai-Pan (Asian Saga  #2)      610         5.0
4  How to Be a Domestic Goddess: Baking and the A...      330         5.0

In [30]:
books_ratings_count = pd.io.sql.read_sql("SELECT title, book_id, author_id, COUNT(rating) AS count_rating, AVG(rating) AS avg_rating FROM books JOIN ratings USING (book_id) GROUP BY (book_id) HAVING COUNT(rating) >= 50 ORDER BY (count_rating) DESC", con = engine)
books_ratings_count

title  book_id  author_id  \
0                             Twilight (Twilight  #1)      948        554   
1                 The Hobbit  or There and Back Again      750        240   
2                              The Catcher in the Rye      673        235   
3                Angels & Demons (Robert Langdon  #1)       75        106   
4   Harry Potter and the Prisoner of Azkaban (Harr...      302        236   
5   Harry Potter and the Chamber of Secrets (Harry...      299        236   
6   Harry Potter and the Order of the Phoenix (Har...      301        236   
7                                         Animal Farm       79        195   
8   The Fellowship of the Ring (The Lord of the Ri...      722        240   
9   Harry Potter and the Half-Blood Prince (Harry ...      300        236   
10                                  Lord of the Flies      405        621   
11                                   Romeo and Juliet      545        630   
12  The Lightning Thief (Percy Jackson and the Oly...      779        498   
13                                    Of Mice and Men      488        311   
14             The Da Vinci Code (Robert Langdon  #2)      696        106   
15                                      The Alchemist      627        469   
16                          The Giver (The Giver  #1)      733        372   
17                                     The Book Thief      656        402   
18                                       Little Women      399        376   

    count_rating  avg_rating  
0            160    3.662500  
1             88    4.125000  
2             86    3.825581  
3             84    3.678571  
4             82    4.414634  
5             80    4.287500  
6             75    4.186667  
7             74    3.729730  
8             74    4.391892  
9             73    4.246575  
10            71    3.901408  
11            66    3.787879  
12            62    4.080645  
13            61    3.622951  
14            59    3.830508  
15            57    3.789474  
16            56    3.750000  
17            53    4.264151  
18            52    4.192308

In [31]:
books_ratings_reviews = pd.io.sql.read_sql("WITH books_ratings_count AS(SELECT title, book_id, author_id, COUNT(rating) AS count_rating, AVG(rating) AS avg_rating FROM books JOIN ratings USING (book_id) GROUP BY (book_id) HAVING COUNT(rating) >= 50 ORDER BY (count_rating) DESC) SELECT author, author_id, AVG(avg_rating) AS avg_author_rating FROM authors JOIN books_ratings_count USING (author_id) GROUP BY (author_id) ORDER BY (avg_author_rating) DESC LIMIT 1", con = engine)
books_ratings_reviews

author  author_id  avg_author_rating
0  J.K. Rowling/Mary GrandPré        236           4.283844

### Conclusion:
- J.K. Rowling/Mary GrandPré has the highest average rating among books with at least 50 ratings.

## Finding the average number of text reviews among users who rated more than 50 books. 

In [32]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [33]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [34]:
reviews_count = pd.io.sql.read_sql("SELECT username, COUNT(review_id) AS reviews_count FROM reviews GROUP BY (username) ORDER BY (reviews_count) DESC", con = engine)
reviews_count.head()

username  reviews_count
0      susan85             29
1  sfitzgerald             28
2   martinadam             27
3  lesliegibbs             26
4       eallen             26

In [35]:
ratings_count = pd.io.sql.read_sql("SELECT username, COUNT(rating_id) AS ratings_count FROM ratings GROUP BY (username) HAVING COUNT(rating_id) > 50 ORDER BY (ratings_count) DESC", con = engine)
ratings_count.head()

username  ratings_count
0          paul88             56
1      martinadam             56
2     sfitzgerald             55
3       richard89             55
4  jennifermiller             53

In [36]:
avg_reviews = pd.io.sql.read_sql("WITH reviews_count AS (SELECT username, COUNT(review_id) AS reviews_count FROM reviews GROUP BY (username)), ratings_count AS (SELECT username, COUNT(rating_id) AS ratings_count FROM ratings GROUP BY (username) HAVING COUNT(rating_id) > 50) SELECT AVG(reviews_count) AS avg_reviews_count FROM ratings_count JOIN reviews_count USING (username)", con = engine)
avg_reviews

avg_reviews_count
0          24.333333

### Conclusion:
- The average number of text reviews among users who rated more than 50 books is around 24.

## Overall Conclusions
- There are 821 books published after January 1, 2020.
- The biggest number of user reviews for a book is 7, and it belongs to Twilligt (#1);
- The highest average rating for a book is 5.0;
- The lowest average rating (1.5) belongs to "Harvesting the Heart";
- Penguin Books has released the greatest number of books (42) of more than 50 pages;
- J.K. Rowling/Mary GrandPré has the highest average rating among books with at least 50 ratings;
- The average number of text reviews among users who rated more than 50 books is around 24.